### ISSUE TITLE
Genie.WebChannels.ChannelMessage doesn't have a defined StructTypes.StructType

### ISSUE THREAD

**Issue by tcarion** [GenieFramework/Genie.jl#397]
**Describe the bug**
I'm getting an `ArgumentError("Genie.WebChannels.ChannelMessage doesn't have a defined StructTypes.StructType")` when using `Genie.WebChannels.broadcast`

**Error stacktrace**
```
ERROR: ArgumentError: Genie.WebChannels.ChannelMessage doesn't have a defined `StructTypes.StructType`
Stacktrace:
 [1] write(::StructTypes.NoStructType, ::Array{UInt8,1}, ::Int64, ::Int64, ::Genie.WebChannels.ChannelMessage; kw::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/tcarion/.julia/packages/JSON3/VnoI3/src/write.jl:91
 [2] write(::StructTypes.NoStructType, ::Array{UInt8,1}, ::Int64, ::Int64, ::Genie.WebChannels.ChannelMessage) at /home/tcarion/.julia/packages/JSON3/VnoI3/src/write.jl:91
 [3] write(::Genie.WebChannels.ChannelMessage; kw::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/tcarion/.julia/packages/JSON3/VnoI3/src/write.jl:39
 [4] write(::Genie.WebChannels.ChannelMessage) at /home/tcarion/.julia/packages/JSON3/VnoI3/src/write.jl:37
 [5] top-level scope at REPL[18]:1
 [6] run_repl(::REPL.AbstractREPL, ::Any; backend_on_current_task::Bool) at /builddir/build/BUILD/julia/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:292
 [7] run_repl(::REPL.AbstractREPL, ::Any) at /builddir/build/BUILD/julia/build/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:288
 [8] (::Base.var"#807#809"{Bool,Bool,Bool,Bool})(::Module) at ./client.jl:399
```
**To reproduce**
I could reproduce the error like this:
```
julia> mess = Genie.WebChannels.ChannelMessage("channel", 0xd58c781368526b14, "message", Dict("payload" => "foo"))
julia> Renderer.Json.JSONParser.json(mess)
```

I would say the issue comes from the migration to JSON3. It worked after I did
```
StructTypes.StructType(::Type{Genie.WebChannels.ChannelMessage}) = StructTypes.Struct()
```

but I admit I don't exactly know what I'm doing here :)

**Additional context**
```
Julia Version 1.5.3
Commit 788b2c77c1* (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Xeon(R) Gold 6254 CPU @ 3.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, cascadelake)
Environment:
  JULIA_REVISE = auto
```
and
```
  [c43c736e] Genie v2.0.4
  [340e8cb6] SearchLight v1.0.2
  [21a827c4] SearchLightSQLite v1.0.0
  [856f2bd8] StructTypes v1.7.2
```

**Comment by essenciary** [comment]
@tcarion Thanks for this! It's definitely related to JSON3 and your solution is most likely the fix :)) I'll take a closer look though.

**Comment by essenciary** [comment]
Added in v3

**Comment by tcarion** [comment]
Hi @essenciary!

I tried with Genie v3, and although I'm not running on error anymore, I still need to do 
´StructTypes.StructType(::Type{Genie.WebChannels.ChannelMessage}) = StructTypes.Struct()´ for the message to be passed correctly to the socket.

To be noted that I'm not using the script provided by `channels_support`, I'm handling the messages clientside with the `rxjs/webSocket` utility of Angular. But the message is still not received by the client anyway.

**Comment by essenciary** [comment]
Thanks. I used the MVE you provided above and worked well. What do you mean by "for the message to be passed correctly to the socket"? What do you do, what do you expect to happen, and what happens?

**Comment by tcarion** [comment]
Sorry I was not precise enough.

For example, if I do `Genie.WebChannels.broadcast("channel", "foo")`, `"foo"` is received by the client as expected (I can see the message in the navigator developper tool).

However, as soon as I add a payload, e.g `Genie.WebChannels.broadcast("channel", "foo", Dict(:payload => "bar"))`, nothing is received clientside anymore.

After I do `StructTypes.StructType(::Type{Genie.WebChannels.ChannelMessage}) = StructTypes.Struct()`, everything works as expected.

**Comment by essenciary** [comment]
Got it, thanks, I'll check it out.

**Comment by essenciary** [comment]
Got it, thanks, I'll check it out.

**Comment by gaelforget** [comment]
I feel that I am having a similar issue in the [Rendering-JSON-views](https://genieframework.github.io/Genie.jl/dev/guides/Working_With_Genie_Apps.html#Rendering-JSON-views) tutorial where I get this error:

```Julia
julia> BooksController.API.billgatesbooks()
ERROR: ArgumentError: BooksController.Book doesn't have a defined `StructTypes.StructType`
Stacktrace:
  [1] write(::StructTypes.NoStructType, buf::Vector{UInt8}, pos::Int64, len::Int64, ::BooksController.Book; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/VnoI3/src/write.jl:91
  [2] write(::StructTypes.NoStructType, buf::Vector{UInt8}, pos::Int64, len::Int64, ::BooksController.Book)
    @ JSON3 ~/.julia/packages/JSON3/VnoI3/src/write.jl:91
  [3] #write#71
```

Out of curiosity, I tried adding the following code snippet in `src/renderers/Json.jl` which does remove the error. However, the resulting `api/v1/bgbooks` page does not look right (just shows _{}_).

```
function JSON3.StructTypes.StructType(::Type{T}) where {T<:DataType}
  JSON3.StructTypes.Struct()
end

function JSON3.StructTypes.StructType(::Type{T}) where {T<:Vector}
  JSON3.StructTypes.Struct()
end
```

Happy to open a separate issue if this is unrelated.

**Comment by essenciary** [comment]
Confirmed, this was an issue. Should now be fixed in Genie v4.1. 

Closing this, please reopen if problem still persists.
